### ISSUE TITLE
Add sub_core_grid support to ttnn.where, ttnn.gt, ttnn.slice

### ISSUE THREAD

**Issue by sraizada-tt** [tenstorrent/tt-metal#32524]
Similar issue for :
typecast op: https://github.com/tenstorrent/tt-metal/pull/21128
unitize op: https://github.com/tenstorrent/tt-metal/pull/16193

**Comment by ekuslysTT** [comment]
@cmaryanTT please assign someone from your team

**Comment by KalaivaniMCW** [comment]
@sraizada-tt  Do you have a preference on which inputs do you need this first ? 
sub_core_grid support for Interleaved inputs first ?

**Comment by sraizada-tt** [comment]
yes, support for DRAM interleaved inputs only would work!

**Comment by cmaryanTT** [comment]
```
# repetition

    # If token appears in prompt or output, apply, otherwise use 1.0 for no-op.
    combined_mask = ttnn.add(context.prompt_mask, context.output_mask)
    penalties = ttnn.where(combined_mask, context.repetition_penalties, 1.0)
    inverse_penalties = ttnn.div(1, penalties)

    # If logits are positive, divide by penalty, otherwise multiply by penalty.
    scaling = ttnn.where(ttnn.gt(logits, 1), inverse_penalties, penalties)

    logits = ttnn.multiply(logits, scaling, output_tensor=logits)

    return logits
```


2 calls here:
ttnn.where(combined_mask, context.repetition_penalties, 1.0)
ttnn.where([ttnn.gt](http://ttnn.gt/)(logits, 1), inverse_penalties, penalties)

what are the shapes of penalties and inverse_penalties?

`[batch_size, vocab_size//devices] so [32, 128*1024] or [32, 128*1024//8]`

**Comment by KalaivaniMCW** [comment]
> yes, support for DRAM interleaved inputs only would work!

Thanks @sraizada-tt 
I'll start with `sub_core_grid` for Ternary infra which would help this case 
`scaling = ttnn.where(ttnn.gt(logits, 1), inverse_penalties, penalties)`

Also I need to recheck if `ttnn.gt(logits, 1)` is using binary or unary infra

**Comment by dchenTT** [comment]
We do not support ttnn.div(1, penalties). Scalar value being the second parameter ttnn.div(penalties, 1) is supported.

**Comment by cmaryanTT** [comment]
@dchenTT - we can do broadcast from 1 as a tensor, right? Does something like ttnn.div([1], penalties) work? Obviously for div order is important

**Comment by sraizada-tt** [comment]
I can do this reciprocal at the torch level I think when creating the tensor -- shouldn't be a problem

**Comment by dchenTT** [comment]
> [@dchenTT](https://github.com/dchenTT) - we can do broadcast from 1 as a tensor, right? Does something like ttnn.div([1], penalties) work? Obviously for div order is important

Yes. The only case we do not support yet is the first operand is a scalar value.

**Comment by dchenTT** [comment]
> > [@dchenTT](https://github.com/dchenTT) - we can do broadcast from 1 as a tensor, right? Does something like ttnn.div([1], penalties) work? Obviously for div order is important
> 
> Yes. The only case we do not support yet is the first operand is a scalar value.

Actually we also have an unary op rdiv can do like this ttnn.rdiv(penalties, 1), but it does not support sub_core_grids yet unless we need to.

**Comment by KalaivaniMCW** [comment]
We are adding sharding support to Ternary infra in #32155 which will also help us add `sub_core_grid` functionality. 
We have tested the functionality locally [here](https://github.com/tenstorrent/tt-metal/compare/mouliraj/ternary_sharding...mouliraj/ternary_sub_core?expand=1) 
Waiting for the above PR to get approvals and go in, so that we can raise this one after CI runs are green

**Comment by KalaivaniMCW** [comment]
PR https://github.com/tenstorrent/tt-metal/pull/32155 adds sharding support and sub_core_grids support to Where TTT/TTS/TST Variants

**Comment by dchenTT** [comment]
PR for basic binary op sub_core_grids is already merged. https://github.com/tenstorrent/tt-metal/pull/32781

**Comment by KalaivaniMCW** [comment]
https://github.com/tenstorrent/tt-metal/pull/32155 was merged 
The TTT/TTS/TST variants of Where op has sub_core_grids support

```
scaling = ttnn.where(ttnn.gt(logits, 1), inverse_penalties, penalties)
penalties = ttnn.where(combined_mask, context.repetition_penalties, 1.0)
```

Do all of these input tensors have the same shape or any broadcast cases involved here ?
ex. Shapes of combined_mask, context.repetition_penalties ?

**Comment by sraizada-tt** [comment]
same shapes

**Comment by sraizada-tt** [comment]
sorry my bad there is a broadcast involved, 
combined_mask is [32, vocab_size]
context.repetition_penalties is [32, 1]

**Comment by sraizada-tt** [comment]
testing these ops:
binary op sub_core_grids works
slice works 

these are failing
1. ttnn.where(combined_mask, context.repetition_penalties, 1.0, **op_kwargs)
2. ttnn.gt(logits, 1,  **op_kwargs)

with Kernel group cores do not match sub device cores for programmable core type TENSIX

**Comment by KalaivaniMCW** [comment]
@sraizada-tt Both of these should be using the binary op infra
Let me check 
cc: @dchenTT

**Comment by KalaivaniMCW** [comment]
I tested these params on main for `ttnn.where(combined_mask, context.repetition_penalties, 1.0 )` bcsat case and it passes

`ttnn_result = ttnn.where(ttnn_C, ttnn_T, ttnn_F, sub_core_grids=sub_core_grid)`

Do you have any specific core configs you'd like me to test  @sraizada-tt ? 
Meanwhile I'll also run this on Debug mode to see if I get any additional info

<img width="1512" height="982" alt="Image" src="https://github.com/user-attachments/assets/bf5ca68c-07a5-41a1-96e1-3663aa022615" />

**Comment by sraizada-tt** [comment]
this is what I am running: 

ttnn_C : Shape([32, 16384]) 
ttnn_T: Shape([32, 1]) 
ttnn_F: 1.0
{'sub_core_grids': {[(x=1,y=0) - (x=3,y=9)], [(x=5,y=0) - (x=6,y=9)]}}

**Comment by sraizada-tt** [comment]
it is because ttnn_C is int32 and ttnn_T is bf16. if I typecast to bf16 it works

**Comment by KalaivaniMCW** [comment]
Oh might be .. there is an implicit conversion for when  ttnn_C is int32 
let me test this out

> could it be because ttnn_C is int32 and ttnn_T is bf16?

**Comment by KalaivaniMCW** [comment]
I checked the core split for bf16 and fp32... let me check this case. 

<img width="974" height="813" alt="Image" src="https://github.com/user-attachments/assets/852255d6-2920-4948-bc3b-dad96deb26b9" />

**Comment by sraizada-tt** [comment]
ok and for ttnn.gt(logits, 1) - logits is Shape([32, 16384]) DataType.BFLOAT8_B  DRAM interleaved

**Comment by KalaivaniMCW** [comment]
Does this fix help resolve `ttnn.where(combined_mask, context.repetition_penalties, 1.0 )`
[ kalaivani/where_int32_typecast ](https://github.com/tenstorrent/tt-metal/tree/kalaivani/where_int32_typecast) ?

**Comment by sraizada-tt** [comment]
if the typecast can't be fused into the op, I prefer to do it at the model level because then we can control deallocates

**Comment by KalaivaniMCW** [comment]
> ok and for ttnn.gt(logits, 1) - logits is Shape([32, 16384]) DataType.BFLOAT8_B DRAM interleaved

Are both the inputs  BFLOAT8_B ? 
I suspect this uses legacy binary ?

**Comment by sraizada-tt** [comment]
what do you mean both the inputs?  ttnn.gt has a tensor input =logits  and a scalar input = 1

**Comment by KalaivaniMCW** [comment]
> if the typecast can't be fused into the op, I prefer to do it at the model level because then we can control deallocates

Oh , Then could you try typecasting the `combined_mask` ( I believe this is the one in INT32 dtype) to bfloat16 in the model and see if the `sub_core_grids` support work  for this use case ?

We haven't introduced mixed dtypes at kernel level yet as this involves a larger infra change.
I'll look at the possibility of adding this atleast for where's condition tensors. 

Meanwhile, let me know if typecasting in the model unblocks you.

**Comment by sraizada-tt** [comment]
yes, typecasting works

**Comment by KalaivaniMCW** [comment]
> yes, typecasting works

Great then. You don't need the fix in [kalaivani/where_int32_typecast ](https://github.com/tenstorrent/tt-metal/tree/kalaivani/where_int32_typecast) right ?

**Comment by sraizada-tt** [comment]
no

**Comment by KalaivaniMCW** [comment]
> ok and for ttnn.gt(logits, 1) - logits is Shape([32, 16384]) DataType.BFLOAT8_B DRAM interleaved

I believe this case also uses typecast to convert  BFLOAT8_B to BFLOAT16 and back
try typecasting the input to bfloat16

**Comment by KalaivaniMCW** [comment]
<img width="1261" height="384" alt="Image" src="https://github.com/user-attachments/assets/ae6e77a3-7284-420b-96f2-bd20b141ea28" />

**Comment by sraizada-tt** [comment]
for bf8 vs bf16 -- can we not fix this in the op at the packing/unpacking ?

**Comment by sraizada-tt** [comment]
typecasting to bf16 makes ttnn.gt work

**Comment by KalaivaniMCW** [comment]
I believe the binary_ng bcast does not work for block formats like bfloat8b/4b .. that's why the inputs are typecasted to bfloat16. 
The legacy binary did offer block_format support broadcast  at kernel level but it works only for broadcasting with Add/mul/sub not ops like GT

cc: @dchenTT I believe that's why we typecast block formats in binary_ng ?

**Comment by sraizada-tt** [comment]
overall, 
all the ops work now @cmaryanTT 
we can close this issue

**Comment by KalaivaniMCW** [comment]
Closing this as discussed

**Comment by dchenTT** [comment]
As from previous comment, binary_ng for ttnn.gt does work, but for block format, it is directed to legacy binary. I am glad to hear it can be worked around.
